File: samples/java/Bingo/Server/Api/src/main/java/systems/zlink/samples/bingo/server/api/BingoPlayerRecordStore.java
package systems.zlink.samples.bingo.server.api;

import org.springframework.stereotype.Component;

import java.util.concurrent.ConcurrentHashMap;

@Component
public final class BingoPlayerRecordStore {
    private final ConcurrentHashMap<String, PlayerRecord> records = new ConcurrentHashMap<>();

    public PlayerRecord get(String actorId) {
        return records.getOrDefault(actorId, new PlayerRecord(actorId, 0, 0));
    }

    public PlayerRecord report(String actorId, boolean won) {
        return records.compute(
                actorId,
                (ignored, current) -> {
                    PlayerRecord record =
                            current == null ? new PlayerRecord(actorId, 0, 0) : current;
                    return won
                            ? new PlayerRecord(actorId, record.wins() + 1, record.losses())
                            : new PlayerRecord(actorId, record.wins(), record.losses() + 1);
                });
    }

    public record PlayerRecord(String actorId, int wins, int losses) {}
}

File: samples/kotlin/Bingo/Server/Api/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/api/BingoPlayerRecordStore.kt
package systems.zlink.samples.kotlin.bingo.server.api

import java.util.concurrent.ConcurrentHashMap
import org.springframework.stereotype.Component

@Component
class BingoPlayerRecordStore {
    private val records = ConcurrentHashMap<String, PlayerRecord>()

    fun get(actorId: String): PlayerRecord = records[actorId] ?: PlayerRecord(actorId, 0, 0)

    fun report(actorId: String, won: Boolean): PlayerRecord =
        records.compute(actorId) { _, current ->
            val record = current ?: PlayerRecord(actorId, 0, 0)
            if (won) record.copy(wins = record.wins + 1)
            else record.copy(losses = record.losses + 1)
        }!!

    data class PlayerRecord(val actorId: String, val wins: Int, val losses: Int)
}
